### PR TITLE
Remove usages of URIUtil.toURI and URIUtil.split

### PR DESCRIPTION
**Pull request by gregw**
Use Resources APIs instead.

This builds on #11568 and completes the move away from explicit URI handling.

### CODE REVIEW

**Comment by joakime** [comment]
This is for 12.0.9, right?

File: jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java [hunk 1]
```diff
@@ -61,22 +60,24 @@ public static URI resolvePathAsLocalizedURI(String path, Bundle bundle, Path jet
         if (StringUtil.isBlank(path))
             return null;
 
-        if (path.startsWith("/") || path.startsWith("file:/")) //absolute location
-            return URIUtil.toURI(path);
-        else
+        if (path.startsWith("file:/"))
+            return new URI(path);
+
+        if (path.startsWith("/") && File.pathSeparatorChar != '/')
```
Comment (joakime) [inline jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java]: Not sure I understand the second condition.
What is that hoping to address?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java [hunk 2]
```diff
@@ -51,6 +52,15 @@ public abstract class Resource implements Iterable<Resource>
     private static final Logger LOG = LoggerFactory.getLogger(Resource.class);
     private static final LinkOption[] NO_FOLLOW_LINKS = new LinkOption[]{LinkOption.NOFOLLOW_LINKS};
 
+    public static final Comparator<Resource> COMPARATOR_BY_NAME = new Comparator<Resource>()
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: Use `ResourceCollators` instead.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java [hunk 3]
```diff
@@ -390,6 +439,62 @@ default Resource newJarFileResource(URI uri)
...
+     *     If the result of processing the input segment is a java archive, then its resulting URI will be a mountable URI as {@code jar:file:...!/}
+     * </p>
+     *
+     * @param str the input string of references
+     */
+    default List<Resource> split(String str)
+    {
+        List<Resource> list = new ArrayList<>();
+
+        StringTokenizer tokenizer = new StringTokenizer(str, ",;|");
```
Comment (joakime) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java:450]: This has traditionally been a static that works well, but in reality it has a few quirks.
Eg: The comma `,` and semicolon `;` are valid characters on all filesystems.

I believe the pipe `|` is also valid on a shocking large number of filesystems (even jar / war / zip too), but don't hold me to that.

Perhaps this needs some javadoc indicating that those characters cannot be used in filename? meh.

File: jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java [hunk 4]
```diff
@@ -189,7 +188,7 @@ public void visitContextParam(WebAppContext context, Descriptor descriptor, XmlP
                 for (String i : values)
                 {
                     String entry = normalizer.expand(i);
-                    tlds.add(URIUtil.toURI(entry).toURL());
+                    tlds.add(context.getResourceFactory().newResource(entry).getURI().toURL());
```
Comment (joakime) [inline jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java:191]: This might create a mount, just to eventually get a URL, throwing away the `Resource` object.

This fact is why `URIUtil.toURI(entry)` existed, to avoid the resource costs.

File: jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java [hunk 5]
```diff
@@ -200,7 +199,8 @@ public void visitContextParam(WebAppContext context, Descriptor descriptor, XmlP
             {
                 List<URI> uris = values.stream()
                     .map(normalizer::expand)
-                    .map(URIUtil::toURI)
+                    .map(context.getResourceFactory()::newResource)
```
Comment (joakime) [inline jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java:202]: Ditto, newResource might create mounts.

**Review by joakime — COMMENTED** [review]
I realize this is a draft, but I'm noticing a few things already.

File: jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java]: It is broken!... it should be just separatorChar.  
The idea is that if the file system uses '/' as the separator, then the Path can deal with it below.  If we see a '/' and the file system is not using '/', then it needs to be handled as a URI.

Comment (gregw) [reply jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java]: @joakime is that better?

File: jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java (hunk 4 above)
Comment (gregw) [reply jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java:191]: Hmmm I'm wondering if we just need `new URI(String)` here.  These should never be relative I think.... pondering....

File: jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java (hunk 5 above)
Comment (gregw) [reply jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java:202]: ditto

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java (hunk 3 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java:450]: I agree it i a bit funny... but because the string can be a list of URIs, if a filename has one of these characters, then it can be listed as an encoded URI.

eitherway... another PR I think...

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 2 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: DOH! I knew something like that existed. Thanks

File: jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java (hunk 4 above)
Comment (gregw) [reply jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java:191]: Actually, I think it is OK.  If the TLDs or other META_INF resources are jar:file URIs, then they will eventually be mounted.   So we will keep the mount point in the context resource factory and we are good!

Comment (joakime) [reply jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java:191]: Note that, on windows, the string `E:/code/base` is a valid String that can be used in Java for Path and File objects.
But as a String, it's seen as a scheme `E` with SSP of `/code/base`.

File: jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java (hunk 1 above)
Comment (joakime) [reply jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java]: This would trip up things on Windows.
Take the input string `E:/code/base` as an example.
The `File.pathSeparatorChar` is `\`.
But that String also conforms to URI space.

```
jshell> var u1 = URI.create("E:/code/base")
u1 ==> E:/code/base

jshell> u1.getScheme()
$2 ==> "E"

jshell> u1.getPath()
$3 ==> "/code/base"

jshell> u1.getSchemeSpecificPart()
$4 ==> "/code/base"
```

And the Windows UNC syntax is equally tricky.

```
jshell> var u1 = URI.create("//machine/share/dir")
u1 ==> //machine/share/dir

jshell> u1.getScheme()
$2 ==> null

jshell> u1.isAbsolute()
$3 ==> false

jshell> u1.getPath()
$4 ==> "/share/dir"

jshell> u1.getAuthority()
$5 ==> "machine"
```

Comment (joakime) [reply jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java]: Those are valid Strings for path references in Windows.
Can be used by `new File(String)` and even `Paths.of(String)` equally successfully.

Comment (gregw) [reply jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java]: @joakime  I don't think think `E:/code/base` is a problem. It doesn't start with '/' so we will try it with `FileSystems.getDefault().getPath(path);`   So I think that is correct.

The line that you highlighted is only if somebody passed a path like `/foo/bar` in.   It starts with '/', but if we are on a file system that also uses '/', then it will be passed to `FileSystems.getDefault().getPath(path);`.  If it is on windows, then the separation char is not '/' so we will try `new URI("file:" + path)`.    So this is only for relative URI paths passed into this method on a windows system.

in short, this line is exactly so this method will work on windows.

(P.S. Not sure why this is not using Resource... but I don't know enough about OSGi resources to change this).

File: jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java (hunk 4 above)
Comment (gregw) [reply jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java:191]: The code in newResource handles that

**Comment by gregw** [comment]
> This is for 12.0.9, right?

I've put in back in 12.0.8 for now... as it is green... but let's discuss at the meeting.

**Comment by joakime** [comment]
This PR has failing jetty-util tests on Windows.

I'm working on collecting the fixes into a different branch ATM.

**Review by joakime — CHANGES_REQUESTED** [review]
Generally like this PR, just some update for Windows needed.
See PR #11585 for those fixes. (merging that PR will update this branch)

**Review by joakime — DISMISSED** [review]

**Comment by gregw** [comment]
@lachlan-roberts, as this include your PR, can I get a review that I have not broken anything you needed?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java [hunk 6]
```diff
@@ -146,19 +151,26 @@ static Resource combine(Resource... resources)
         return CombinedResource.combine(List.of(resources));
     }
 
+    static Resource resolveElseNew(Resource base, String pathOrUri, Function<String, Resource> factory)
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java]: Javadoc?

**Review by lorban — DISMISSED** [review]
Just one nit about that `ResourceFactory.resolveElseNew()` function lacking javadoc; otherwise LGTM.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java (hunk 6 above)
Comment (gregw) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java]: It was only used once. So I inlined it instead

**Review by lorban — APPROVED** [review]

**Comment by hengsin** [comment]
hi, the WebInfConfiguration.unpack(WebAppContext context) changes of this ticket is not being make to the ee8/ee9 source, is that as intended or just being overlook ? We runs into a Windows only problem (deployment of a OSGi WAB jar file) that seems due to the missing of this change in the ee8/ee9 source.

**Comment by joakime** [comment]
@hengsin please open a new issue, and provide details on what is happening for your scenario. (your comment here doesn't have enough details to action off of)
